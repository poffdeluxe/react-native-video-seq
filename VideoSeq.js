import React from 'react';
import PropTypes from 'prop-types';
import Video from 'react-native-video';

class VideoSeq extends React.Component {
  constructor(props) {
    super(props);

    this.onEnd = this.onEnd.bind(this);

    this.state = {
      curVidIndex: 0,
    };
  }

  onEnd(event) {
    // Call user defined onEnd func if it exists
    if (this.props.onEnd) {
      this.props.onEnd(event);
    }

    // Switch to the next video in the sequence
    let curVidIndex = this.state.curVidIndex + 1;
    if(curVidIndex == this.props.videos.length) {
      // Reset to first vid if repeat is on
      curVidIndex = this.props.repeat ? 0 : this.state.curVidIndex;
    }

    this.setState({curVidIndex});
  }

  render() {
    const curVid = this.props.videos[this.state.curVidIndex];

    const { source, repeat, onEnd, ...passThroughProps } = this.props;
    passThroughProps.onEnd = this.onEnd;

    return (
      <Video source={curVid} {...passThroughProps} />
    )
  }
}

VideoSeq.propTypes = {
  videos: PropTypes.array.isRequired,
  ...Video.propTypes
};

export default VideoSeq;
