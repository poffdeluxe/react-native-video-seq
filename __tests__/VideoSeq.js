import 'react-native';
import React from 'react';
import VideoSeq from '../VideoSeq';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

// Mock out the video stuff
jest.mock('react-native-video', () => 'Video');

it('changes to the next video when one ends', () => {
  const videos = [
    'vid1.mov',
    'vid2.webm',
    'vid3.avi'
  ];

  const component = renderer.create(
    <VideoSeq videos={videos} />
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.getInstance().onEnd();
  expect(component.toJSON()).toMatchSnapshot();

  component.getInstance().onEnd();
  expect(component.toJSON()).toMatchSnapshot();

  // Should stay on vid3.avi since repeat is not on
  component.getInstance().onEnd();
  expect(component.toJSON()).toMatchSnapshot();
});

it('resets to the first vid if repeat is on', () => {
  const videos = [
    'vid1.mov',
    'vid2.webm'
  ];

  const component = renderer.create(
    <VideoSeq videos={videos} repeat={true} />
  );

  expect(component.toJSON()).toMatchSnapshot();

  // Should start vid2
  component.getInstance().onEnd();
  expect(component.toJSON()).toMatchSnapshot();

  // Should repeat back to vid1
  component.getInstance().onEnd();
  expect(component.toJSON()).toMatchSnapshot();
});

it('calls passed in onEnd function if it exists', () => {
  const videos = [
    'vid1.mov'
  ];

  const mockCallback = jest.fn();

  const component = renderer.create(
    <VideoSeq videos={videos} repeat={true} onEnd={mockCallback} />
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.getInstance().onEnd();
  expect(component.toJSON()).toMatchSnapshot();

  expect(mockCallback.mock.calls.length).toBe(1);
});
