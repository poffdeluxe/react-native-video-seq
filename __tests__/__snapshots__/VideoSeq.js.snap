// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls passed in onEnd function if it exists 1`] = `
<Video
  onEnd={[Function]}
  source="vid1.mov"
  videos={
    Array [
      "vid1.mov",
    ]
  }
/>
`;

exports[`calls passed in onEnd function if it exists 2`] = `
<Video
  onEnd={[Function]}
  source="vid1.mov"
  videos={
    Array [
      "vid1.mov",
    ]
  }
/>
`;

exports[`changes to the next video when one ends 1`] = `
<Video
  onEnd={[Function]}
  source="vid1.mov"
  videos={
    Array [
      "vid1.mov",
      "vid2.webm",
      "vid3.avi",
    ]
  }
/>
`;

exports[`changes to the next video when one ends 2`] = `
<Video
  onEnd={[Function]}
  source="vid2.webm"
  videos={
    Array [
      "vid1.mov",
      "vid2.webm",
      "vid3.avi",
    ]
  }
/>
`;

exports[`changes to the next video when one ends 3`] = `
<Video
  onEnd={[Function]}
  source="vid3.avi"
  videos={
    Array [
      "vid1.mov",
      "vid2.webm",
      "vid3.avi",
    ]
  }
/>
`;

exports[`changes to the next video when one ends 4`] = `
<Video
  onEnd={[Function]}
  source="vid3.avi"
  videos={
    Array [
      "vid1.mov",
      "vid2.webm",
      "vid3.avi",
    ]
  }
/>
`;

exports[`resets to the first vid if repeat is on 1`] = `
<Video
  onEnd={[Function]}
  source="vid1.mov"
  videos={
    Array [
      "vid1.mov",
      "vid2.webm",
    ]
  }
/>
`;

exports[`resets to the first vid if repeat is on 2`] = `
<Video
  onEnd={[Function]}
  source="vid2.webm"
  videos={
    Array [
      "vid1.mov",
      "vid2.webm",
    ]
  }
/>
`;

exports[`resets to the first vid if repeat is on 3`] = `
<Video
  onEnd={[Function]}
  source="vid1.mov"
  videos={
    Array [
      "vid1.mov",
      "vid2.webm",
    ]
  }
/>
`;
